Fix #4582: Comments after end of line

alanmcruickshank opened the pull request:
This fixes #4582.

Multiline comments would obscure the closing comment of a line. I did a little bit of tidy up here too to mitigate the potential performance implications of this change.

Also:
- Precalculate `IndentStats` on instantiation to simplify logic later on (and probably also get a performance boost).
- Moved the bracket check that was is `get_indent_impulse` into `_crawl_indent_points` because that's the only function that was actually using the results.

WittierDinosaur reviewed (approved):
LGTM - your description says you're removing the ANSI hack, but you're not. I'll approve and leave unmerged, as not sure if you want to fix the ANSI hack first

alanmcruickshank commented:
> LGTM - your description says you're removing the ANSI hack, but you're not. I'll approve and leave unmerged, as not sure if you want to fix the ANSI hack first

Ah - good point. I moved that to #4762 instead because it was causing trouble. I've removed the note from this PR about that.
